web/js/codeworld_shared.js

```diff
@@ -1203,4 +1203,19 @@ function printMessage(type, message) {
     }
 
     outputDiv.scrollTop = outputDiv.scrollHeight;
+
+    if (isRuntime && type === "error") {
+        parseCompileErrors(message).forEach(
+            (error) => {
+                runtimeErrors[error.fullText] = error;
+            }
+        )
+        let cm = window.codeworldEditor;
+        cm.performLint();
+    }
+}
+
+function matchWarningToDeferredError(msg){
+    return msg.replace("warning", "error")
+        + "\n(deferred type error)";
```

Comment:
I didn't realize that GHC adds this "(deferred type error)" thing.  I think we should strip it off, since it won't be meaningful to students who will just come to expect the new behavior.

I do wonder, though, whether we ought to also promote these deferred errors to full errors in the UI.  I mean, maybe the program should still run, but we should display them as errors from the start?  In that case, we could look for "(deferred type error)", strip it off, but also change "warning" to "error" at the same time.

If you do it that way, then there's no need to go back later and fix up the errors when the program crashes.  You'd just filter out the runtime version of the deferred error, or maybe replace it with a brief message like "Giving up because of the error at program.hs:10:42".

Reply:
Ok, got it.

Reply:
There is some confusion. When it is warning it have no "(deferred type error)" postfix, only when it is already runtime error. So I have no quick way to understand if this warning is future error.

Reply:
Okay, I see that now.  It looks like the only way to recognize these deferred warnings is actually before the error sanitizer runs.  Here's what I get from GHC itself:

    Test.hs:2:21: warning: [-Wdeferred-type-errors]
        • No instance for (Fractional Int) arising from the literal ‘4.5’
        • In the first argument of ‘take’, namely ‘4.5’
          In the second argument of ‘($)’, namely ‘take 4.5 "hello"’
          In the expression: print $ take 4.5 "hello"

or

    Test.hs:2:21: warning: [-Wdeferred-out-of-scope-variables]
        Variable not in scope: n :: Int

However, those brackets are stripped off by the error sanitizer at https://github.com/google/codeworld/blob/master/codeworld-error-sanitizer/src/ErrorSanitizer.hs#L28 before the message even leaves the server.

I think one could write a rule in ErrorSanitizer.hs to replace "warning: (.\*) \\[-Wdefer-.\*\\]" with "error: \1" on the server.  This would just change the text of the message so the client would render it as an error, but it wouldn't cause the compile to actually fail.  That might have the effect we want.

Reply:
Looks like haskell version of [replace function](https://github.com/google/codeworld/blob/master/codeworld-error-sanitizer/src/RegexShim.hs#L50) does not support group regex replacement.
Can I implement this?

Reply:
Ah, yes.  I think it would be better to use `subRegex` at http://hackage.haskell.org/package/regex-compat-0.95.1/docs/Text-Regex.html

Reply:
After another look, it seems that depends on regex-posix, which isn't good.  Among other things, it can't handle unicode characters, which is definitely something we're using now. :(  So, maybe that's just a reference you could use for a new implementation of replacement on top of regex-tdfa?

Reply:
Ok, I happy to try implement this. Anyway in this case we don't need full regex support with nested groups. I'll start from simple replacement without nesting, and will look how it going.

Reply:
Sure, whatever works! :)  Just make sure that whatever part you implement is consistent between the two implementations in RegexShim